[Cypress-e2e] Add feast smoke tests

Srihari1192 opened the pull request:
<!--- If this is a non-code change, this template is not required; reference any issues or top-level descriptions as needed -->
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- https://issues.redhat.com/browse/RHOAIENG-123456 -->
https://issues.redhat.com/browse/RHOAIENG-33423
## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->

- Added Feature Store Smoke Tests 
- Added Supporting method createFeatureStoreCR got feature store CR creation

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tested in local and CI against RHOAI Nightly build
## Test Impact
<!--- What tests have you done to cover the implemented functionality -->
<!--- If tests are not applicable, explain why here -->
None - this is a test

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)
- [ ] The code follows our [Best Practices](/docs/best-practices.md) (React coding standards, PatternFly usage, performance considerations)

If you have UI changes: 
<!--- You can ignore these if you are doing manifest, backend, internal logic, etc changes; aka non-UI / visual changes -->
- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change.

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

Srihari1192 commented:
@antowaddle @manosnoam @ConorOM1 Please review it 
Tested this changes in CI `/job/dashboard-tests/2964/` but tests failing because of few development changes needs to cherry-pick to `stable-2.x` branch

Srihari1192 commented on frontend/src/__tests__/cypress/cypress/utils/oc_commands/featureStoreResources.ts:
```diff
@@ -0,0 +1,39 @@
+    // Apply the modified YAML
+    applyOpenShiftYaml(yamlContent);
+    //wait for the feature store cr to be created
+    waitForPodReady('feast-test-s3', '300s', namespace);
```
Sure

Srihari1192 commented on frontend/src/__tests__/cypress/cypress/tests/e2e/featureStore/testFeatureStoreEmptyState.cy.ts:
```diff
@@ -0,0 +1,38 @@
+import { retryableBefore } from '#~/__tests__/cypress/cypress/utils/retryableHooks';
+import type { FeatureStoreTestData } from '#~/__tests__/cypress/cypress/types';
+
+describe('Verify Feature Store Default page Contents', () => {
```
Done

Srihari1192 commented:
> Thanks @Srihari1192 for this contribution, I've added some general comments, can you please implement them generally to your tests? Also, can you change the merge target to main rather than the stable branch?
> 
> Before these test merge, we need a clean run on an ODH/RHOAI cluster, will that be possible?

@antowaddle I updated the PR with changes 
As main branch contains 3.0 changes so added to stable-2.x branch as this changes supposed to required for 2.25 testing 
I tested here `job/dashboard-tests/2996/` , test looks good

antowaddle commented:
> > Thanks @Srihari1192 for this contribution, I've added some general comments, can you please implement them generally to your tests? Also, can you change the merge target to main rather than the stable branch?
> > Before these test merge, we need a clean run on an ODH/RHOAI cluster, will that be possible?
> 
> @antowaddle I updated the PR with changes As main branch contains 3.0 changes so added to stable-2.x branch as this changes supposed to required for 2.25 testing I tested here `job/dashboard-tests/2996/` , test looks good

@Srihari1192 please open this PR against `main` first - see https://redhat-internal.slack.com/archives/C05SMJ09DD2/p1758299575741169 You can then cherry pick the changes here accordingly into the stable-2.x branch.

Srihari1192 commented:
> > > Thanks @Srihari1192 for this contribution, I've added some general comments, can you please implement them generally to your tests? Also, can you change the merge target to main rather than the stable branch?
> > > Before these test merge, we need a clean run on an ODH/RHOAI cluster, will that be possible?
> > 
> > 
> > @antowaddle I updated the PR with changes As main branch contains 3.0 changes so added to stable-2.x branch as this changes supposed to required for 2.25 testing I tested here `job/dashboard-tests/2996/` , test looks good
> 
> @Srihari1192 please open this PR against `main` first - see https://redhat-internal.slack.com/archives/C05SMJ09DD2/p1758299575741169 You can then cherry pick the changes here accordingly into the stable-2.x branch. We can then continue the review there, thank you.

oh ok I see that some conflict changes with new nav items in [featureStoreGlobal.ts]( https://github.com/opendatahub-io/odh-dashboard/blob/main/frontend/src/__tests__/cypress/cypress/pages/featureStore/featureStoreGlobal.ts#L100 ) is it ok to update according to 2.25

antowaddle commented:
> featureStoreGlobal.ts
@Srihari1192 I would suggest aligning with what is in featureStoreGlobal.ts for main and also for 2.25 where applicable. Otherwise, your're going to see Mock/e2e Test failures/ As per that thread, if this is 3.0 only and should merge into main then I would suggest to review this with the team.

antowaddle commented on frontend/src/__tests__/cypress/cypress/tests/e2e/featureStore/testFeatureStoreSmokeTests.cy.ts:
```diff
@@ -0,0 +1,113 @@
+        featureStoreGlobal.navigateToEntities();
+        featureStoreGlobal.selectProject(testData.feastCreditScoringProject);
+        featureStoreGlobal.shouldHavePageDescription();
+        featureEntitiesTable.shouldHaveEntityCount(4);
```
Is there a better way to do this? What happens with an additional entity is added, will this fail? Could we retrieve the entity details via OC and then validate them directly?

antowaddle commented on frontend/src/__tests__/cypress/cypress/tests/e2e/featureStore/testFeatureStoreSmokeTests.cy.ts:
```diff
@@ -0,0 +1,113 @@
+        featureStoreGlobal.navigateToDatasets();
+        featureStoreGlobal.shouldHavePageDescription();
+        featureDataSetsTable.findTable().should('be.visible');
+        featureDataSetsTable.shouldHaveDataSetCount(9);
```
Same comment as the entity line above

antowaddle commented on frontend/src/__tests__/cypress/cypress/tests/e2e/featureStore/testFeatureStoreSmokeTests.cy.ts:
```diff
@@ -0,0 +1,113 @@
+        cy.step('Verify Data Sources Page');
+        featureStoreGlobal.navigateToDataSources();
+        featureStoreGlobal.shouldHavePageDescription();
+        featureDataSourcesTable.shouldHaveDataSourceCount(5);
```
Same comment as the entity line above

antowaddle commented on frontend/src/__tests__/cypress/cypress/tests/e2e/featureStore/testFeatureStoreSmokeTests.cy.ts:
```diff
@@ -0,0 +1,113 @@
+        cy.step('Verify Features Page');
+        featureStoreGlobal.navigateToFeatures();
+        featureStoreGlobal.shouldHavePageDescription();
+        featuresTable.shouldHaveFeatureCount(10);
```
Same comment as the entity line above

antowaddle commented on frontend/src/__tests__/cypress/cypress/tests/e2e/featureStore/testFeatureStoreSmokeTests.cy.ts:
```diff
@@ -0,0 +1,113 @@
+        featureStoreGlobal.navigateToFeatureServices();
+        featureStoreGlobal.shouldHavePageDescription();
+        featureServicesTable.shouldHaveFeatureServiceCount(9);
+      } finally {
```
We don't typically use 'finally' in our framework, can you use retryableBefore and after here please?

antowaddle reviewed (changes requested):
Hey @Srihari1192 I did another review, there are some changes required here.

antowaddle commented on frontend/src/__tests__/cypress/cypress/tests/e2e/featureStore/testFeatureStorePageContents.cy.ts:
```diff
@@ -0,0 +1,129 @@
+
+        // Create route and get all counts from API
+        createRoute(projectName, testData.feastInstanceName);
+        cy.get<string>('@routeUrl').then((routeUrl) => {
```
If this API call fails, we should see the response, error code etc. Can you add that also please?

antowaddle reviewed (commented):
@Srihari1192 I've added a few additional comments.

In a more general sense, the tests written here are maybe better suited to a Mock test flow, they are fine to keep for basic sanity but we typically exercise full e2e user flows with our Cypress e2e tests (see https://github.com/opendatahub-io/odh-dashboard/tree/main/frontend/src/__tests__/cypress/cypress/tests/e2e - 'Writing Tests').

Srihari1192 commented:
> @Srihari1192 I've added a few additional comments.
> 
> In a more general sense, the tests written here are maybe better suited to a Mock test flow, they are fine to keep for basic sanity but we typically exercise full e2e user flows with our Cypress e2e tests (see https://github.com/opendatahub-io/odh-dashboard/tree/main/frontend/src/__tests__/cypress/cypress/tests/e2e - 'Writing Tests').

Yes agree added these basic scenarios to make sure feature store smoke validation, will add other tests with e2e user flow
